rotkehlchen/chain/mixins/rpc_nodes.py

```diff
@@ -0,0 +1,374 @@
+import logging
+import random
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Sequence
+from contextlib import suppress
+from typing import TYPE_CHECKING, Any, Generic, TypeVar
+from urllib.parse import urlparse
+
+import requests
+from ens import ENS
+from httpx import HTTPError
+from solana.rpc.api import Client
+from solana.rpc.core import RPCException
+from typing_extensions import NamedTuple
+from web3 import HTTPProvider, Web3
+from web3.exceptions import Web3Exception
+from web3.middleware import ExtraDataToPOAMiddleware
+
+from rotkehlchen.chain.ethereum.constants import ETHEREUM_ETHERSCAN_NODE
+from rotkehlchen.chain.evm.types import NodeName, WeightedNode
+from rotkehlchen.constants.misc import ONE
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.fval import FVal
+from rotkehlchen.greenlets.manager import GreenletManager
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.tests.utils.globaldb import Literal
+from rotkehlchen.types import (
+    SUPPORTED_CHAIN_IDS,
+    SUPPORTED_EVM_CHAINS_TYPE,
+    ChainID,
+    ChecksumEvmAddress,
+    EVMTxHash,
+    SupportedBlockchain,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+WEB3_NODE_TYPE = TypeVar('WEB3_NODE_TYPE', Web3, Client)
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Web3Node(NamedTuple, Generic[WEB3_NODE_TYPE]):
+    """This represents an RPC node with its capabilities."""
+    web3_instance: WEB3_NODE_TYPE
+    is_pruned: bool
+    is_archive: bool
+
+
+class RPCManagerMixin(ABC, Generic[WEB3_NODE_TYPE]):
+    """
+    Mixin that provides logic for managing RPC nodes. It tracks active connections
+    and implements the core mechanisms for connecting to them.
+
+    This mixin is intended for chains that support RPCs, such as EVM-based chains
+    and Solana. It is not used directly, since each chain introduces small
```

Comment:
Can't we do the same for Substrate chains? (not in this PR ofc)

Reply:
yes. The idea I had with this is that the core logic that is querying from the db and managing we can do for any chain and can be abstracted